### DISCUSSION TITLE
How to query relationship with `order_by()` and `limit()` ?

### DISCUSSION THREAD

**Discussion by ikus060** [sqlalchemy/sqlalchemy#10209] [Usage Questions] [answered]
Hello,

Thanks in advance for helping me.

To centralize the business logic, I'm using relationship in my ORM model to create relation between objects. Most of the time it suits my need and the query generate is working fine. But, I have a scenario, where loading all the object of the relationship would not make sens. I need to limit the number of object return by the relationship and I'm not sure how to use the API to do that.

I found reference to `with_loader_criteria()`. But it doesn't seams possible to control the OrderBy and Limit.

Here a snippet to demonstrate my problem.

```
from sqlalchemy import Column, Integer, String, ForeignKey
from sqlalchemy.orm import Session, relationship, remote, foreign, declarative_base
from sqlalchemy import create_engine, event
from sqlalchemy.engine import Engine

Base = declarative_base()


@event.listens_for(Engine, 'connect')
def enable_sqlite_foreign_keys(connection, connection_record):
    """
    Enable Foreign key for SQLite
    """
    if 'sqlite3' in str(connection.__class__):
        cursor = connection.cursor()
        cursor.execute("PRAGMA foreign_keys=ON;")
        cursor.close()


class Person(Base):
    __tablename__ = 'person'
    id = Column(Integer(), primary_key=True)
    name = Column(String(80))


class Group(Base):
    __tablename__ = 'group'
    letter = Column(String(), primary_key=True)
    persons = relationship(
        Person, primaryjoin=remote(foreign(Person.name)).startswith(letter), lazy=True, viewonly=True
    )


e = create_engine("sqlite://", echo=True)
Base.metadata.create_all(e)

session = Session(e)

group_f = Group(letter='F')
group_k = Group(letter='K')
frank = Person(name='Frank')
fred = Person(name='Fred')
fany = Person(name='Fany')
francine = Person(name='Francine')
fabio = Person(name='Fabio')
felix = Person(name='Felix')
katy = Person(name='Katy')
session.add_all([group_f, group_k, frank, fred, fany, francine, fabio, felix, katy])
session.commit()

# SELECT "group".letter AS group_letter FROM "group"  WHERE "group".letter = ?
session.expire(group_f)
assert group_f.letter == 'F'

# Using the property is working
# The query generate is perfect too.
# SELECT person.id AS person_id, person.name AS person_name FROM person WHERE (person.name LIKE ? || '%')
assert len(group_f.persons) == 6

# But What if I want to list the persons order by name and limit to 5 ?
# SELECT person.id AS person_id, person.name AS person_name FROM person WHERE (person.name LIKE ? || '%') ORDER BY person.name LIMIT 5
session.query(group_f.persons).order_by(Person.name).limit(5).all()

```

**Comment by zzzeek** [comment, marked as answer]
"limit" is not intrinsic to relationship loaders since there are many styles of relationship loader that are not compatible with LIMIT without adding more complex subquery logic.    if you need relationships that use LIMIT, that is normally done structurally with window functions, as in the example you see at https://docs.sqlalchemy.org/en/20/orm/join_conditions.html#row-limited-relationships-with-window-functions  .

You could use a relatoinship as the above and then the LIMIT portion of it would be controlled by the WHERE clause which you can then add using [relationship criteria](https://docs.sqlalchemy.org/en/20/orm/queryguide/relationships.html#adding-criteria-to-loader-options).   

for ORDER BY, if you are using built-in loaders, order by currently has to be  part of the relationship itself using the [order_by](https://docs.sqlalchemy.org/en/20/orm/relationship_api.html#sqlalchemy.orm.relationship.params.order_by) parameter.


Otherwise, the other way that people might have a relationship be represented by completely custom logic is to use the [contains_eager](https://docs.sqlalchemy.org/en/20/orm/queryguide/relationships.html#contains-eager) approach where you write a [LEFT OUTER] JOIN that contains what you need, then route that into the collection.

**Reply by jd-solanki** [reply]
Hi @zzzeek 

This is now common scenario for people using FastAPI, You can also check the demo example [here](https://sqlmodel.tiangolo.com/tutorial/relationship-attributes/read-relationships/#get-a-list-of-relationship-objects) in SQLModel docs.

Is it possible for SQLAlchemy to provide nice abstraction over limit & order by to easily loading related data? I'm beginner in SQLAlchemy and its really hard for me to understand what's going on.

Thanks for such powerful library ❤️

**Reply by zzzeek** [reply]
I can explain to you exactly what's going on at all times.  We are using Python objects to construct SQL statement strings, which we send to a database to get results.     So it is your job to always be thinking in terms of SQL queries you want to emit and results you want to get back, as though you were writing the application without the benefit of SQLAlchemy automating those tasks.   

in this case, we are talking about a SELECT with LIMIT we want to emit.  I would disregard SQLModels' docs as well as the very exotic use case discussed on this issue and simply do the thing you want, emit SELECT with LIMIT:

```py
stmt = select(TheThingIWant).where(.... some filtering criteria....).limit(10)
objects = session.scalars(stmt).all()
```

note there's kind of a "step 2. :  ???" in there. how do we filter to get related objects?  There's a helper you can use called [with_parent](https://docs.sqlalchemy.org/en/20/orm/queryguide/api.html#sqlalchemy.orm.with_parent).  Put the parent object there.   here's a demo

```py
from __future__ import annotations

from sqlalchemy import create_engine
from sqlalchemy import ForeignKey
from sqlalchemy import select
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column
from sqlalchemy.orm import relationship
from sqlalchemy.orm import Session
from sqlalchemy.orm import with_parent


class Base(DeclarativeBase):
    pass


class A(Base):
    __tablename__ = "a"

    id: Mapped[int] = mapped_column(primary_key=True)
    bs: Mapped[list[B]] = relationship("B")


class B(Base):
    __tablename__ = "b"
    id: Mapped[int] = mapped_column(primary_key=True)
    a_id: Mapped[int] = mapped_column(ForeignKey("a.id"))


e = create_engine("sqlite://", echo=True)
Base.metadata.create_all(e)

s = Session(e)

s.add(A(bs=[B(), B(), B(), B(), B()]))
s.commit()


my_a = s.get(A, 1)

three_bs = s.scalars(select(B).where(with_parent(my_a, A.bs)).limit(3)).all()

print(three_bs)
```

**Reply by CaselIT** [reply]
there is also https://docs.sqlalchemy.org/en/20/orm/large_collections.html#write-only-relationships that instead of a list fives you a select you can manipulate as you please:
https://docs.sqlalchemy.org/en/20/orm/large_collections.html#querying-items

**Reply by jd-solanki** [reply]
Thanks for clarification 🙏🏻 

Just to make sure we're on same path. Whatever API we use there'll always be two DB queries, Right?
1. For getting parent object
2. Will get items for that parent object with our own filter & order_by

**Reply by CaselIT** [reply]
In the general case yes.
If you are selecting a single model and you have only a single one-to-many relationship you may also use a manual join with contains_eager. something like
```py
select(A).joins(A.bs).options(contains_eager(A.bs)).limit(3).where(A.id==the_id)
```
